modules/account-lib/src/coin/sol/tokenEncodeDecode.ts

```diff
@@ -0,0 +1,216 @@
+import { AccountMeta, PublicKey, Signer, TransactionInstruction } from '@solana/web3.js';
+const splToken = require('@solana/spl-token');
+
+export enum TokenInstruction {
+  InitializeMint = 0,
+  InitializeAccount = 1,
+  InitializeMultisig = 2,
+  Transfer = 3,
+  Approve = 4,
+  Revoke = 5,
+  SetAuthority = 6,
+  MintTo = 7,
+  Burn = 8,
+  CloseAccount = 9,
+  FreezeAccount = 10,
+  ThawAccount = 11,
+  TransferChecked = 12,
+  ApproveChecked = 13,
+  MintToChecked = 14,
+  BurnChecked = 15,
+  InitializeAccount2 = 16,
+  SyncNative = 17,
+  InitializeAccount3 = 18,
+  InitializeMultisig2 = 19,
+  InitializeMint2 = 20,
+}
+
+export abstract class TokenError extends Error {
+  constructor(message?: string) {
+    super(message);
+  }
+}
+
+/** Thrown if an instruction's program is invalid */
+export class TokenInvalidInstructionProgramError extends TokenError {
+  name = 'TokenInvalidInstructionProgramError';
+}
+
+/** Thrown if an instruction's keys are invalid */
+export class TokenInvalidInstructionKeysError extends TokenError {
+  name = 'TokenInvalidInstructionKeysError';
+}
+
+/** Thrown if an instruction's type is invalid */
+export class TokenInvalidInstructionTypeError extends TokenError {
+  name = 'TokenInvalidInstructionTypeError';
+}
+
+/** TODO: docs */
+export interface TransferCheckedInstructionData {
+  instruction: TokenInstruction.TransferChecked;
+  amount: number;
+  decimals: number;
+}
+
+export function addSigners(keys: AccountMeta[], ownerOrAuthority: PublicKey, multiSigners: Signer[]): AccountMeta[] {
+  if (multiSigners.length) {
+    keys.push({ pubkey: ownerOrAuthority, isSigner: false, isWritable: false });
+    for (const signer of multiSigners) {
+      keys.push({ pubkey: signer.publicKey, isSigner: true, isWritable: false });
+    }
+  } else {
+    keys.push({ pubkey: ownerOrAuthority, isSigner: true, isWritable: false });
+  }
+  return keys;
+}
+
+/**
+ * Construct a TransferChecked instruction
+ *
+ * @param source       Source account
+ * @param mint         Mint account
+ * @param destination  Destination account
+ * @param owner        Owner of the source account
+ * @param amount       Number of tokens to transfer
+ * @param decimals     Number of decimals in transfer amount
+ * @param multiSigners Signing accounts if `owner` is a multisig
+ * @param programId    SPL Token program account
+ *
+ * @return Instruction to add to a transaction
+ */
+export function createTransferCheckedInstruction(
```

Comment:
double check if we need this. if we can create a transfer checked instruction in the current, stable version of the token library, we should use it.

Reply:
The existing createTransferCheckedInstruction relies on a structure that causes the errors in alpha (https://github.com/solana-labs/solana-program-library/blob/87fe3793ea87aac5630564c86c1b7792bb3e8ace/token/ts/src/instructions/transferChecked.ts#L22)

Creating a new one allows us to work around this struct error by encoding/decoding through JSON